core/network/impl/router_libp2p.hpp

```diff
@@ -99,12 +102,39 @@ namespace kagome::network {
       });
     }
 
-    /**
-     * Process a received Propagate Transactions message
-     */
-    bool processPropagateTransactionsMessage(
-        const libp2p::peer::PeerId &peer_id,
-        const PropagatedTransactions &msg) const;
+    template <typename T, typename H, typename F>
+    void readAsyncMsgWithHandshake(std::shared_ptr<Stream> stream,
+                                   H &&handshake,
+                                   F &&f) const {
+      auto read_writer = std::make_shared<ScaleMessageReadWriter>(stream);
+      read_writer->read<std::decay_t<H>>([stream,
+                                          handshake{std::move(handshake)},
+                                          read_writer,
+                                          wself{weak_from_this()},
+                                          f{std::forward<F>(f)}](
+                                             auto &&read_res) mutable {
+        if (!read_res) {
+          if (auto self = wself.lock())
+            self->log_->error("Error while reading handshake: {}",
```

Comment:
Shouldn't it be reported when self is null either?

Reply:
No, it's an async operation and can be theoreticaly call when application began to destruct.